docs/admin-guide/kubernetes-in-production.rst

```diff
@@ -0,0 +1,160 @@
+
+* SSL termination by ALB: convenient usage of ACM (free Amazon CA) and certificates upload to AWS IAM
+* Using the “new” ELBv2 Application Load Balancer
+
+We use Skipper_ as our HTTP proxy to route based on Host header and path. Skipper directly comes with a Kubernetes data client to automatically update its routes periodically.
+
+Mate_ is automatically configuring the Ingress hosts as DNS records in Route53 for us.
+
+.. _Kube AWS Ingress Controller: https://github.com/zalando-incubator/kube-ingress-aws-controller
+.. _Skipper: https://github.com/zalando/skipper
+.. _Mate: https://github.com/zalando-incubator/mate
+
+Resources
+=========
+
+Understanding the Kubernetes resource requests and limits is crucial.
+
+Default resource requests and limits can be configured via the LimitRange_ resource. This can prevent “stupid” incidents like JVM deployments without any settings (no memory limit and no JVM heap set) eating all the node’s memory.
+
+We provide a `tiny script`_ and use the Downwards API to conveniently run JVM applications on Kubernetes without the need to manually set the maximum heap size.
+
+Kubelet can be instructed to reserve a certain amount of resources for the system and for Kubernetes components (kubelet itself and Docker etc). Reserved resources are subtracted from the `node’s allocatable resources`_. This improves scheduling and makes resource allocation/usage more transparent. Node allocatable resources or rather reserved resources are also visible in `Kubernetes Operational View`_.
+
+*TODO: add link to kubelet flags*
+
+.. _LimitRange: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/admission_control_limit_range.md
+.. _tiny script: https://github.com/zalando/docker-openjdk/blob/master/utils/java-dynamic-memory-opts
+.. _node’s allocatable resources: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node-allocatable.md
+
+Graceful Pod Termination
+========================
+
+Kubernetes will cause service disruptions on pod terminations by default as applications and configuration need to be prepared for graceful shutdown. By default, pods receive the TERM signal and will be killed 30s later with KILL. Kubernetes expects the container to handle the TERM signal and change the readinessProbe_ to “fail”. This can be achieved by changing an in-memory value which instructs the health endpoint to return a non-200 status code. Sadly the default probe settings (period 10s, failure threshold 3) prevent an in-time change of the “ready” state. To achieve graceful pod shutdown..
+
+* ..the container needs to switch its health endpoint to non-200 status code when receiving the TERM signal
+* ..the readinessProbe settings need to be tweaked to react within the grace period (30s by default), e.g. by setting failure threshold to 1 or 2 (while keeping the default period of 10s)
+
+TODO: example and link existing blog posts by others
```

Comment:
This part is not correct. I described the problem/tested it in another comment on our internal issue tracker:

> So I wasn't convinced that Kubernetes still sends requests after the SIGTERM. I did some tests, and I'm pretty sure that the requests you will see after the SIGTERM are caused by keep-alive connections.
> 
> If you disabled keep-alive when receiving the SIGTERM, you will only get a few requests after (until kube-proxy updates), but the readinessProbe has no effect after the SIGTERM.
Looking at the logs presented in https://blog.risingstack.com/graceful-shutdown-node-js-kubernetes/ actually also shows this, he stops getting requests before failureThreshold * periodSecond
> 
> Here is my test program and results (logs): https://github.com/mikkeloscar/kube-sigterm-test